### PR TITLE
Replace `String.format(…, args)` with `"…".formatted(args)`

### PR DESCRIPTION
**Pull request by jxblum**
See #2751.

Please read the description in #2751.

I also think formatted Strings work nicer if natural line breaks are preserved.

For example:

```java
throw new IllegalArgumentException("some message with placeholders"
    .formatted(arg1, arg2, ..., argN);
```

Rather than this non-sense:

```java
throw new IllegalArgumentException("some message with placeholders".formatted(arg1, arg2, ...,
    argN);
```

That is orphans, or even partly broken argument lists.

Finally...

This is not strictly necessary. But, in my opinion, I think this is a nice improvement over our existing String formatting.

I also took the opportunity to clean up a (minimal) few compiler warnings while stumbling across message formatting.

Also, I changed a couple assertions that unnecessarily used a `Supplier` that did not need to and added `Suppliers` to a few assertions that should have had them.

### CODE REVIEW

File: src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java [hunk 1]
```diff
@@ -223,7 +223,7 @@ public int getDatabase() {
 	@Override
 	public void setDatabase(int index) {
 
-		Assert.isTrue(index >= 0, () -> String.format("Invalid DB index '%d'; non-negative index required", index));
```
Comment (kdebski85) [inline src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java:226]: I do not think this is a good change.
Without a supplier, we are needlessly constructing a string.
This assertion is almost always true, so it should be optimised for a happy path.
The same applies to other similar changes.

Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java:226]: Thank you for the feedback. I agree.

I wrote simple [JMH Benchmark](https://github.com/codeprimate-software/cp-labs/blob/main/src/main/java/org/cp/labs/java/SupplierCreationVsStringFormattingBenchmark.java#L49) today comparing `String.formatted(..)` with and without the use of a `Supplier` to the assertion, such as `isTrue(..)` above. It turns out that, without the `Supplier` the invocation performance (throughput & average time) is much worse over a large number of iterations, even despite the added Object allocations for `Suppliers`.  I did not look at the Heap memory map, but I suspect it will create a a bit more garbage than necessary, at least until JIT compiler optimizations kick in.

However, I will point out that, I was careful to only remove the `Suppliers` in non-critical code paths, such as in this case... "Configuration".

Configuration is only really ever in invoked on Spring container startup, once, when the beans are constructed, configured and initialized.

Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java:226]: I forget to mention, keep in mind that the `Supplier` Lambda in this case is NOT a non-capturing Lambda since it uses arguments from the enclosing method. Therefore, it, too, results in Object construction every time the method is called (though `RedisSentinelConfiguration.setDatabase(:int)` is only likely to be called once in this case).

Still, let's be clear here what we are talking about in terms of "optimizations".

You must always consider resource utilization when considering performance as often they are related. You must also measure. And finally, you must take into consideration the code path. Sacrificing readability, correctness or other factors for the sake of performance is not always the best decision.

Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java:226]: In reviewing my changes for this PR, along with writing a JMH Benchmark, I definitely think the use of a `Supplier` in assertions is necessary in SD Redis command classes, where the code path will be hotter at runtime than it will in configuration classes, for instance, such as [here](https://github.com/spring-projects/spring-data-redis/pull/2752/files#diff-868b06883aad7a48f4b3409cc46a399488849084497a79f2ca5db8fdb9d1b0d4R443), as well as in similar places.

In other cases, such as the [src/main/java/org/springframework/data/redis/core/types/RedisClientInfo.java](https://github.com/spring-projects/spring-data-redis/pull/2752/files#diff-41752625b4904fa3640910749360a5ee416a64a34958e026c24dcd1247cd6445) and [BinaryKeyspaceIdentifier](https://github.com/spring-projects/spring-data-redis/pull/2752/files#diff-e19cfb83e92d69907bb21398465ac8d6dfce633cb29d3b8b6e2ea0ef05ba9169R1301), I introduced the use of a `Supplier` inside the assertion where a `Supplier` was not used before.

Comment (jxblum) [reply src/main/java/org/springframework/data/redis/connection/RedisSentinelConfiguration.java:226]: I have updated this PR to (re)introduce the use of the `Supplier` inside assertions. Only very few were inside critical codepaths, like command classes (specifically Jedis and Lettuce ZSet commands). The rest were in configuration classes (non-critical codepaths).

It should be noted, that most of the String formatting occurs inside Exception construction (e.g. `InvalidDataAccessApiUsageException`) and Logger statements. There is no option for using a `Supplier` in most of these cases, not without extra guards; for example:

```java
public void someMethod(Object... args) {
  //...
    logDebug(() -> 
        "Some formatted String with args [%s]".formatted(Arrays.toString(args));
}

private void logDebug(Supplier<String> logMessage) {
    if (logger.isDebugEnabled()) {
        logger.debug(logMessage.get());  
    }
}
```

This may make for some nice optimizations down the road (and in certain cases, I noticed I have introduced log helper methods, just like described above), but that is not a concern in this PR.

In conclusion, I still feel the appearance of using `String.formatted(..)` is much nicer than `Spring.format(..)`, as it calls attention to the Exception/Log/Whatever message immediately.

**Comment by mp911de** [comment]
Thank you for your contribution. That's merged, polished, and backported now.
